sdk/src/client/worker.rs

```diff
@@ -0,0 +1,239 @@
+    Receive(#[from] RecvError),
+    #[error("error exiting worker: {0}")]
+    Join(#[from] JoinError),
+}
+
+#[derive(Debug)]
+pub(crate) struct WorkerPool(Mutex<VecDeque<Arc<Worker>>>);
+
+impl WorkerPool {
+    pub(crate) fn new(count: usize) -> Self {
+        let mut workers = VecDeque::with_capacity(count);
+        for _ in 0..count {
+            workers.push_back(Arc::new(Worker::spawn()));
+        }
+        Self(Mutex::new(workers))
+    }
+
+    pub(crate) async fn process_task<F: 'static + Future + Send>(
+        &self,
+        priority: TaskPriority,
+        future: F,
+    ) -> Result<F::Output, WorkerError>
+    where
+        F::Output: Send,
+    {
+        let mut pool = self.0.lock().await;
+        let worker = pool.front().ok_or(WorkerError::EmptyPool)?.clone();
+        // Move the worker to the back
+        pool.rotate_left(1);
+        drop(pool);
+        let output = worker.process_task(priority, future).await?;
+        Ok(output)
+    }
+
+    pub(crate) async fn resize(&self, new_size: usize) -> Result<(), WorkerError> {
+        if new_size == 0 {
+            return Err(WorkerError::EmptyPool);
+        }
+        let mut pool = self.0.lock().await;
+        let curr_size = pool.len();
+        match new_size.cmp(&curr_size) {
+            core::cmp::Ordering::Less => {
+                while pool.len() > new_size {
+                    if let Some(worker) = pool.pop_front() {
+                        worker.exit()?;
+                    }
+                }
+            }
+            core::cmp::Ordering::Greater => {
+                while pool.len() < new_size {
+                    pool.push_front(Arc::new(Worker::spawn()));
+                }
+            }
+            core::cmp::Ordering::Equal => (),
+        }
+        Ok(())
+    }
+
+    pub(crate) async fn size(&self) -> usize {
+        self.0.lock().await.len()
+    }
+}
+
+pub(crate) enum WorkerEvent {
+    Task(WorkerTask),
+    Exit,
+}
+
+#[repr(u8)]
+#[derive(PartialEq, Eq, PartialOrd, Ord, Debug, Copy, Clone)]
+pub enum TaskPriority {
+    Low = 0,
+    Medium = 1,
+    High = 2,
+}
+
+#[derive(Debug)]
+pub(crate) struct Worker {
+    join_handle: JoinHandle<()>,
+    sender: tokio::sync::mpsc::UnboundedSender<WorkerEvent>,
+}
+
+impl Worker {
+    fn spawn() -> Self {
+        let (sender, mut recv) = tokio::sync::mpsc::unbounded_channel();
+        let join_handle = tokio::spawn(async move {
+            let mut queue = BinaryHeap::new();
+            let mut exiting = false;
+            // Wait to be awakened by the channel
+            while let Some(task) = recv.recv().await {
+                match task {
+                    WorkerEvent::Task(task) => queue.push(task),
+                    WorkerEvent::Exit => exiting = true,
+                }
+                loop {
+                    if !exiting {
+                        // Get up to 10 messages at a time
+                        for _ in 0..10 {
```

Comment:
Why 10?

Reply:
But this will send 10 requests sequentially, only if more than 10 requests are waiting, it will be done in parallel by using another worker, right? Just having a single message seems to be faster

Reply:
No, it will receive 10 messages from the incoming requests, and add those to the priority queue. Then it will process them sequentially by priority.

Reply:
Also, the workers do not pull from the same messages, they cannot steal messages from another worker.

Reply:
Well that's what I mean, if one worker has 10 messages, and no other worker anything, then it will send these 10 requests sequentially
If every worker would only have a single message, then the other 9 messages would be split over other workers and could all be sent at the same time (as long as they don't exceed the limit)?

Reply:
yes, but that cannot happen because the work is distributed across the workers evenly

Reply:
It is still possible for a worker to be assigned high cost work and thereby end up with a backlog though. If they shared an input source then this could not happen, but then you have to deal with lock contention and I'm not sure it's worth that.

Reply:
Yeah you're right probably not worth it